### ISSUE TITLE
✨ OpenSearch description for all pages

### ISSUE THREAD

**Issue by APT37** [neon-mmd/websurfx#622]
### Description

Link an [OpenSearch description](https://developer.mozilla.org/en-US/docs/Web/OpenSearch) to all pages so browsers can offer adding the the websurfx instance as a search engine.

### Do you want to work on this issue?

Yes

**Comment by neon-mmd** [comment]
@APT37 
Sorry for delay in the reply.

Thanks ❤️ for taking the time to open this issue, this really helps improve the project, so we really appreciate it 👍 .

Since you had requested to work on this issue, so we have assigned the issue to you, we hope to see you soon with a cool and wonderful PR. 🚀

**Comment by neon-mmd** [comment]
@APT37 
Sorry for the ping.

Also, we would like to ask like why would we need an open search description for all the pages? I mean it seems quite unreasonable because this is a search engine and to be honest, we should not only have open search description for only the search page? Moreover, the search page will only be the one to be grabbed by the web browsers anyway. So would like to the know the reason behind adding it for all the pages though? 🤔

**Comment by APT37** [comment]
Should we change it to mean only the homepage then?

**Comment by neon-mmd** [comment]
> Should we change it to mean only the homepage then?

Nah, I think it should only be the search page (where all the search results will be displayed). 😅

**Comment by APT37** [comment]
So that's what you meant. But why not the homepage, too?

**Comment by neon-mmd** [comment]
> So that's what you meant. But why not the homepage, too?

Yes because the web browsers grab the search page of the search engine you are adding and not there homepage, additionally, web browsers only grab the search page because it is the page that will become your search engine in your search bar and browsers will naturally add the query in the format which looks like this `https://search_engine_name.tld/search?q=<user search query>`, so you can understand why they would naturally prefer grabbing the search page of the search engine rather then the homepage. 😅

**Comment by APT37** [comment]
As I understand it, browsers either use the search fields or the given OpenSearch descriptor file - the latter taking precedence. Having such a file on both the homepage and the search result page makes this more uniform, hence my proposal to add it to those pages at least.

**Comment by neon-mmd** [comment]
> As I understand it, browsers either use the search fields or the given OpenSearch descriptor file - the latter taking precedence. Having such a file on both the homepage and the search result page makes this more unform, hence my proposal to add it to those pages at least.

Ok sure, then let's go for it. 👍

**Comment by APT37** [comment]
Just noticed this pretty much depends on https://github.com/neon-mmd/websurfx/issues/589.

```xml
-- websurfx.xml --

<OpenSearchDescription
  xmlns="http://a9.com/-/spec/opensearch/1.1/"
  xmlns:moz="http://www.mozilla.org/2006/browser/search/">
  <ShortName>websurfx</ShortName>
  <Description>websurfx is an open source alternative to searx, which provides a modern-looking, lightning-fast, privacy respecting, secure meta search engine</Description>
  <InputEncoding>UTF-8</InputEncoding>
  <Image width="16" height="16" type="image/x-icon">https://websurfx.example.org/favicon.ico</Image>
  <Url type="text/html" template="[searchURL]"/>
</OpenSearchDescription>
```

```html
-- link to put on the home and search pages --

<link
  rel="search"
  type="application/opensearchdescription+xml"
  title="websurfx"
  href="https://websurfx.example.org/websurfx.xml" />
```

**Comment by neon-mmd** [comment]
> Just noticed this pretty much depends on [#589](https://github.com/neon-mmd/websurfx/issues/589).
> 
> -- websurfx.xml --
> 
> <OpenSearchDescription
>   xmlns="http://a9.com/-/spec/opensearch/1.1/"
>   xmlns:moz="http://www.mozilla.org/2006/browser/search/">
>   <ShortName>websurfx</ShortName>
>   <Description>websurfx is an open source alternative to searx, which provides a modern-looking, lightning-fast, privacy respecting, secure meta search engine</Description>
>   <InputEncoding>UTF-8</InputEncoding>
>   <Image width="16" height="16" type="image/x-icon">https://websurfx.example.org/favicon.ico</Image>
>   <Url type="text/html" template="[searchURL]"/>
> </OpenSearchDescription>
> -- link to put on the home and search pages --
> 
> <link
>   rel="search"
>   type="application/opensearchdescription+xml"
>   title="websurfx"
>   href="https://websurfx.example.org/websurfx.xml" />

@APT37 
Sorry for the delay in the response. But if I am not wrong, you can omit the need for favicons for now, that is you can remove the line requiring favicons completely and the file will still be functional.

**Comment by neon-mmd** [comment]
@APT37 
Sorry for the ping, any updates on this issue? 😅

**Comment by neon-mmd** [comment]
@APT37 
Sorry for the ping, any updates on this issue? 😅 

---
The cruise ship needs a docking area. Wait you are working on one right? 😉

**Comment by neon-mmd** [comment]
@APT37 
Sorry for the ping, but since you have not been replying as well as not working on the issue. So I have decided to de-assign you from the issue and take up this work. I hope you don't mind it.

**Comment by neon-mmd** [comment]
## Update
I opened the [issue](https://github.com/actix/actix-web/issues/4038) at `actix-web` for the bug about the custom content-type header being overridden by actix-web for the opensearch description file being served.
